proposals/0000-local-import.rst

```diff
@@ -0,0 +1,198 @@
+
+To avoid polluting the namespace with conflicting identifiers, Haskell programmers can choose between explicit and qualified imports ; however, they are often reluctant to do so as:
+
+- Explicit import lists are inconvenient (they must be edited whenever the code changes). 
+- Module qualifiers can severely affect readability.
+
+The latter problem is often addressed by choosing one-letter qualifiers, which can only be understood by looking up their definition at the top of the file. It is an imperfect solution at best, and does not work well for operators ; for this reason, operators are frequently overloaded via lawless typeclasses for mere syntactic convenience.
+
+OCaml 3.12, released in 2011, introduced two syntactic contructs for locally opening modules: ``let open Module in …`` and ``Module.( … )``. In less than 10 years, this feature has gained widespread adoption throughout the ecosystem, bringing significant benefits in terms of readability and maintainability. For the reasons stated above, we believe Haskell should follow its example.
+
+Proposed Change Specification
+-----------------------------
+We allow qualified and non-qualified ``import`` statements at the **beginning** of any ``let``/``where``/``do``-block.
+
+For all constructs, the effect of such imports is limited to the scope of the enclosing block. For ``let`` and ``where``, the imports also affect the corresponding expression. For example:
+::
+
+  let import Foo in <expr>
+allows the use of symbols from ``Foo`` in the ``<expr>``. Similary, the same thing is possible with:
+::
+
+  f = <expr>
+    where import Foo
+The set of visible typeclass instances at any given point is the union of instances defined in modules imported in all enclosing scopes.  
+
+Moreover, ``import`` statements are allowed to refer to any module qualifier specified outside that block (that is, local ``import`` statements are not limited to full module names). This means that the following is allowed:
+::
+
+  import qualified Data.Set as Set
+  foo = …
+    where import Set
+          …
+As well as:
+::
+
+  foo = …
+    where import qualified Data.Set as Set
+          bar = …
+            where import Set
+                  …
+But the following is not valid:
+::
+
+  foo = …
+    where import qualified Data.Set as Set
+          import Set
+This is consistent with today's semantics for ``import``, which does not allow for:
+::
+
+  module MyModule where
+
+  import qualified Data.Set as Set
+  import Set
+
+As is already the case, module qualifiers can be re-used. For example, in:
+::
+
+  import Foo as A
+
+  main :: IO ()
+  main = do
+    import Bar as A
+    A.x
+
+The symbol ``x`` in ``A.x`` is searched in both ``Bar`` and ``Foo``.
+
+Finally, we allow the syntactic shortcut ``Qualifier.{ <expression> }``, which simply desugars to:
```

Comment:
I think that, while convenient, I'd be slightly opposed to this aspect of the proposal since it introduces brand new syntax. I wonder about the plausibility of using e.g. quasiquoters here? Local imports would increase the utility of this ticket being implemented: https://gitlab.haskell.org/ghc/ghc/issues/1475

Reply:
QuasiQuoters are definitely a good idea. I have mentioned the idea in the alternatives. Thanks !